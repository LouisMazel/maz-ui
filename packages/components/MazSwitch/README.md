# MazSwitch

## Props

<!-- @vuese:MazSwitch:props:start -->

| Name     | Description | Type      | Required | Default |
| -------- | ----------- | --------- | -------- | ------- |
| value    | -           | `Boolean` | `true`   | -       |
| disabled | -           | `Boolean` | `false`  | false   |
| id       | -           | `String`  | `false`  | null    |
| name     | -           | `String`  | `false`  | null    |
| color    | -           | `String`  | `false`  | primary |

<!-- @vuese:MazSwitch:props:end -->

## Events

<!-- @vuese:MazSwitch:events:start -->

| Event Name | Description            | Parameters |
| ---------- | ---------------------- | ---------- |
| input      | Return the input value | Boolean    |

<!-- @vuese:MazSwitch:events:end -->

## MixIns

<!-- @vuese:MazSwitch:mixIns:start -->

| MixIn    |
| -------- |
| uniqueId |

<!-- @vuese:MazSwitch:mixIns:end -->
