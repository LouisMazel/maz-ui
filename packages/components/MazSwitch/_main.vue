<template>
  <div class="maz-switch">
    <input
      :id="uniqueId"
      v-bind="$attrs"
      type="checkbox"
      :name="name"
      :checked="value"
      :disabled="disabled"
      class="maz-switch__input"
      @change="$emit('input', $event.target.checked)"
    >
    <label
      :for="uniqueId"
      class="maz-switch__toggle"
      :class="[bgColorClassTransparency]"
    >
      <span :class="[bgColorClass]" />
    </label>
  </div>
</template>

<script>
import uniqueId from './../../mixins/uniqueId'

export default {
  name: 'MazSwitch',
  mixins: [uniqueId],
  props: {
    value: { type: Boolean, required: true },
    disabled: { type: Boolean, default: false },
    id: { type: String, default: null },
    name: { type: String, default: null },
    color: { type: String, default: 'primary' }
  },
  computed: {
    bgColorClassTransparency () {
      return this.value ? `maz-bg-${this.color}-before-transparency` : 'maz-bg-grey-before-transparency'
    },
    bgColorClass () {
      return this.value ? `maz-bg-${this.color}` : 'maz-bg-white'
    }
  }
}
</script>
