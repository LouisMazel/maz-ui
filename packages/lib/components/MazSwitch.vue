<script lang="ts" setup>
import { type HTMLAttributes, computed } from 'vue'
import { useInstanceUniqId } from '../modules/composables/use-instance-uniq-id'
import type { Color } from './types'

export type { Color }

defineOptions({
  inheritAttrs: false,
})

const props = withDefaults(defineProps<Props>(), {
  style: undefined,
  class: undefined,
  modelValue: false,
  id: undefined,
  disabled: false,
  name: undefined,
  color: 'primary',
})

const emits = defineEmits<{
  /**
   * Emitted when the model value changes
   * @property {boolean} value - The new value
   */
  (event: 'update:model-value', value: boolean): void
  /**
   * Emitted when the model value changes
   * @property {boolean} value - The new value
   */
  (event: 'change', value: boolean): void
}>()

export interface Props {
  /** Style attribut of the component root element */
  style?: HTMLAttributes['style']
  /** Class attribut of the component root element */
  class?: HTMLAttributes['class']
  /** The model value of the switch */
  modelValue?: boolean
  /** The id of the switch */
  id?: string
  /** If the switch is disabled */
  disabled?: boolean
  /** The name of the switch */
  name?: string
  /** Text label */
  label?: string
  /** The color of the switch */
  color?: Color
  /** Whether there is an error with the input. */
  error?: boolean
  /** Whether the input is successful. */
  success?: boolean
  /** Whether there is a warning with the input. */
  warning?: boolean
  /** The hint text to display below the input. */
  hint?: string
}

const instanceId = useInstanceUniqId({
  componentName: 'MazCheckbox',
  providedId: props.id,
})

const bgColorClassVar = computed(() => `var(--maz-color-${props.color}-alpha)`)

const bgColorStyle = computed<HTMLAttributes['class']>(() =>
  props.modelValue ? `var(--maz-color-${props.color})` : 'var(--maz-color-white)',
)

function emit(e: Event) {
  const target = e.target as HTMLInputElement | undefined
  emits('update:model-value', target?.checked ?? false)
  emits('change', target?.checked ?? false)
}
</script>

<template>
  <label
    :for="instanceId"
    class="m-switch"
    :class="[{ '--is-disabled': disabled }, props.class]"
    role="switch"
    :style="style"
    :aria-checked="modelValue"
    tabindex="0"
  >
    <input
      :id="instanceId"
      v-bind="$attrs"
      type="checkbox"
      :name="name"
      :checked="modelValue"
      :aria-label="label"
      :disabled="disabled"
      class="m-switch__input"
      @change="emit"
    >
    <span class="m-switch__toggle" />

    <span v-if="$slots.default || label || hint" class="m-switch__text">
      <!--
        @slot The label of the switch
          @binding {Boolean} value - The value of the switch
      -->
      <slot :value="modelValue">
        {{ label }}
      </slot>

      <span
        v-if="hint"
        class="m-switch__hint" :class="{
          '--error': error,
          '--success': success,
          '--warning': warning,
        }"
      >{{ hint }}</span>
    </span>
  </label>
</template>

<style lang="postcss">
  .m-switch {
  @apply maz-relative maz-inline-flex maz-cursor-pointer maz-items-center maz-gap-2 maz-align-top;

  &.--is-disabled {
    @apply maz-cursor-not-allowed;
  }

  &__input {
    @apply maz-absolute;

    left: -9999px;
  }

  &__toggle {
    @apply maz-h-6 maz-w-12 maz-relative;

    &::before {
      content: '';
      transition: all 200ms ease-in-out;

      @apply maz-relative maz-left-1 maz-top-1 maz-block maz-h-4 maz-w-10 maz-rounded-full;

      background-color: v-bind('bgColorClassVar');
    }

    &::after {
      content: '';

      @apply maz-absolute maz-left-0 maz-top-0 maz-block maz-h-6 maz-w-6 maz-rounded-full;

      background-color: v-bind('bgColorStyle');
      box-shadow: 0 2px 8px 0 hsl(0deg 0% 0% / 20%);
      transition: all 200ms ease-in-out;
    }
  }

  &__input:checked + .m-switch__toggle {
    &::after {
      @apply maz-translate-x-6;
    }
  }

  &__input:disabled {
    + .m-switch__toggle {
      &::after {
        @apply maz-bg-color-light dark:maz-bg-color-lighter;
      }

      &::before {
        @apply maz-bg-color-lighter dark:maz-bg-color-light;
      }
    }
  }

  &__text {
    @apply maz-flex maz-flex-col maz-gap-0;
  }

  &__hint {
    @apply maz-text-sm maz-text-muted;

    &.--error {
      @apply maz-text-danger-600;
    }

    &.--success {
      @apply maz-text-success-600;
    }

    &.--warning {
      @apply maz-text-warning-600;
    }
  }
}
</style>
