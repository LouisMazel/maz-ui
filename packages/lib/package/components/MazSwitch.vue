<template>
  <div class="m-switch" :class="[`--${color}`]">
    <input
      :id="uniqueId"
      v-bind="$attrs"
      type="checkbox"
      :name="name"
      :checked="modelValue"
      :disabled="disabled"
      class="m-switch__input"
      @change="emit"
    />
    <label
      :for="uniqueId"
      class="m-switch__toggle"
      :style="[{ '--m-switch-bg-bar': bgColorClassVar } as StyleValue]"
    >
      <span :style="[bgColorStyle]" />
    </label>
  </div>
</template>

<script lang="ts">
  export type { Color } from './types'
</script>

<script lang="ts" setup>
  import {
    computed,
    getCurrentInstance,
    type PropType,
    type StyleValue,
  } from 'vue'
  import type { Color } from './types'

  const props = defineProps({
    modelValue: { type: Boolean, required: true },
    id: { type: String, default: undefined },
    disabled: { type: Boolean, default: false },
    name: { type: String, default: undefined },
    color: { type: String as PropType<Color>, default: 'primary' },
  })

  const emits = defineEmits(['update:model-value'])

  const uniqueId = computed(
    () => props.id ?? `mazSwitch-${getCurrentInstance()?.uid}`,
  )

  const bgColorClassVar = computed(
    () => `var(--maz-color-${props.color}-alpha)`,
  )

  const bgColorStyle = computed<StyleValue>(() => ({
    backgroundColor: props.modelValue
      ? `var(--maz-color-${props.color})`
      : 'var(--maz-color-white)',
  }))

  const emit = (e: Event) => {
    const target = e.target as HTMLInputElement | undefined
    emits('update:model-value', target?.checked ?? false)
  }
</script>

<style lang="postcss">
  .m-switch {
    &__input {
      @apply maz-absolute;

      left: -9999px;
    }

    &__toggle {
      @apply maz-relative maz-block maz-h-6 maz-w-12 maz-cursor-pointer;

      transform: translate3d(0, 0, 0);

      &::before {
        content: '';
        transition: all 200ms ease-in-out;

        @apply maz-relative maz-top-1 maz-left-1 maz-block maz-h-4 maz-w-10 maz-rounded-full;

        background-color: var(--m-switch-bg-bar);
      }

      & span {
        @apply maz-absolute maz-top-0 maz-left-0 maz-block maz-h-6 maz-w-6 maz-rounded-full;

        box-shadow: 0 2px 8px 0 rgb(0 0 0 / 20%);
        transition: all 200ms ease-in-out;
      }
    }

    &__input:checked + &__toggle {
      span {
        transform: translateX(1.5em);

        &::before {
          transform: scale(1);
          opacity: 0;
          transition: all 400ms ease-in-out;
        }
      }
    }

    &__input:disabled + &__toggle {
      @apply maz-cursor-not-allowed;
    }
  }
</style>
