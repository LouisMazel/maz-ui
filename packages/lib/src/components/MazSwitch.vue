<script lang="ts" setup>
import type { HTMLAttributes } from 'vue'
import type { MazColor } from './types'
import { computed, ref } from 'vue'

import { useInstanceUniqId } from '../composables/useInstanceUniqId'

defineOptions({
  inheritAttrs: false,
})

const props = withDefaults(defineProps<MazSwitchProps>(), {
  style: undefined,
  class: undefined,
  modelValue: false,
  id: undefined,
  disabled: false,
  name: undefined,
  color: 'primary',
})

const emits = defineEmits<{
  /**
   * Emitted when the model value changes
   * @property {boolean} value - The new value
   */
  'update:model-value': [value: boolean]
  /**
   * Emitted when the model value changes
   * @property {boolean} value - The new value
   */
  'change': [value: boolean]
  /**
   * Emitted when the switch lost focus
   * @property {FocusEvent} value - The focus event
   */
  'blur': [value: FocusEvent]
  /**
   * Emitted when the switch is focused
   * @property {FocusEvent} value - The focus event
   */
  'focus': [value: FocusEvent]
}>()

export interface MazSwitchProps {
  /** Style attribut of the component root element */
  style?: HTMLAttributes['style']
  /** Class attribut of the component root element */
  class?: HTMLAttributes['class']
  /** The model value of the switch */
  modelValue?: boolean
  /** The id of the switch */
  id?: string
  /** If the switch is disabled */
  disabled?: boolean
  /** The name of the switch */
  name?: string
  /** Text label */
  label?: string
  /** The color of the switch */
  color?: MazColor
  /** Whether there is an error with the input. */
  error?: boolean
  /** Whether the input is successful. */
  success?: boolean
  /** Whether there is a warning with the input. */
  warning?: boolean
  /** The hint text to display below the input. */
  hint?: string
}

const instanceId = useInstanceUniqId({
  componentName: 'MazSwitch',
  providedId: props.id,
})

const bgColorClassVar = computed(() => {
  return `hsl(var(--maz-${props.color}))`
})

function emit() {
  emits('update:model-value', !props.modelValue)
  emits('change', !props.modelValue)
}

const inputRef = ref<HTMLInputElement>()

function keyboardHandler(event: KeyboardEvent) {
  if (['Space'].includes(event.code)) {
    event.preventDefault()
    emit()
  }
}

function onBlur(event: FocusEvent) {
  inputRef.value?.dispatchEvent(new Event('blur'))
  emits('blur', event)
}
function onFocus(event: FocusEvent) {
  inputRef.value?.dispatchEvent(new Event('focus'))
  emits('focus', event)
}
</script>

<template>
  <label
    :for="instanceId"
    class="m-switch m-reset-css"
    :class="[{ '--is-disabled': disabled }, props.class]"
    role="switch"
    :style="[style, { '--switch-color': bgColorClassVar }]"
    :aria-checked="modelValue"
    tabindex="0"
    @blur="onBlur"
    @focus="onFocus"
    @keydown="keyboardHandler"
  >
    <input
      :id="instanceId"
      v-bind="$attrs"
      ref="inputRef"
      type="checkbox"
      :name="name"
      tabindex="-1"
      :checked="modelValue"
      :aria-label="label"
      :disabled="disabled"
      class="m-switch__input"
      @change="emit"
    >
    <span class="m-switch__toggle" />

    <span v-if="$slots.default || label || hint" class="m-switch__text">
      <!--
        @slot The label of the switch
          @binding {Boolean} value - The value of the switch
      -->
      <slot :value="modelValue">
        {{ label }}
      </slot>

      <span
        v-if="hint"
        class="m-switch__hint" :class="{
          '--error': error,
          '--success': success,
          '--warning': warning,
        }"
      >{{ hint }}</span>
    </span>
  </label>
</template>

<style lang="postcss">
  .m-switch {
  @apply maz-relative maz-inline-flex maz-cursor-pointer maz-items-center maz-gap-2 maz-align-top;

  &:has(input:disabled) {
    @apply maz-cursor-not-allowed;
  }

  &__input {
    @apply maz-absolute;

    left: -9999px;
  }

  &__toggle {
    @apply maz-h-6 maz-w-12 maz-relative;

    &::before {
      content: '';
      transition: all 200ms ease-in-out;

      @apply maz-relative maz-left-0 maz-top-0.5 maz-block maz-h-6 maz-w-[3rem] maz-rounded-full;
      @apply maz-bg-surface-600 dark:maz-bg-surface-400 maz-border maz-border-solid maz-border-divider;
    }

    &::after {
      content: '';

      @apply maz-absolute maz-left-0.5 maz-top-1 maz-block maz-h-5 maz-w-5 maz-rounded-full maz-bg-surface;

      box-shadow: 0 0 4px 0 hsl(0deg 0% 0% / 20%);
      transition: all 200ms ease-in-out;
    }
  }

  &__input:checked + .m-switch__toggle {
    &::after {
      @apply maz-translate-x-6;
    }

    &::before {
      background-color: var(--switch-color);
    }
  }

  &__input:disabled {
    + .m-switch__toggle {
      &::before {
        @apply maz-bg-surface-600 dark:maz-bg-surface-400;
      }

      &::after {
        @apply maz-bg-surface-700 dark:maz-bg-surface-300;

        box-shadow: none;
      }
    }
  }

  &__text {
    @apply maz-flex maz-flex-col maz-gap-0;
  }

  &__hint {
    @apply maz-text-sm maz-text-muted;

    &.--error {
      @apply maz-text-destructive-600;
    }

    &.--success {
      @apply maz-text-success-600;
    }

    &.--warning {
      @apply maz-text-warning-600;
    }
  }
}
</style>
