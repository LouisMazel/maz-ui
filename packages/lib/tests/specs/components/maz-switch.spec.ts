import { mount, type VueWrapper } from '@vue/test-utils'
import MazSwitch from '@components/MazSwitch.vue'

describe('MazSwitch', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    wrapper = mount(MazSwitch)
  })

  test('Should match with the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a checkbox input', () => {
    const wrapper = mount(MazSwitch, {
      props: {
        modelValue: false,
      },
    })

    const checkboxInput = wrapper.find('input[type="checkbox"]')
    expect(checkboxInput.exists()).toBe(true)
  })

  it('emits the "update:model-value" event when the checkbox is clicked', async () => {
    const wrapper = mount(MazSwitch, {
      props: {
        modelValue: false,
      },
    })

    const checkboxInput = wrapper.find('input[type="checkbox"]')
    await checkboxInput.setChecked(true)

    expect(wrapper.emitted('update:model-value')).toBeTruthy()
    expect(wrapper.emitted('update:model-value')[0]).toEqual([true])
  })

  it('applies the "disabled" class when the "disabled" prop is set to true', () => {
    const wrapper = mount(MazSwitch, {
      props: {
        modelValue: false,
        disabled: true,
      },
    })

    expect(wrapper.classes()).toContain('--is-disabled')
  })

  it('toggles the switch when the checkbox is clicked', async () => {
    const wrapper = mount(MazSwitch, {
      props: {
        modelValue: true,
      },
    })

    expect(wrapper.attributes('aria-checked')).toEqual('true')

    await wrapper.setProps({ modelValue: false })
    expect(wrapper.attributes('aria-checked')).toBe('false')
  })
})
