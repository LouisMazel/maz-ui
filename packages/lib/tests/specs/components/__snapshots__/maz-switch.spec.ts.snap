// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MazSwitch > Should match with the snapshot 1`] = `
"<label for=\\"MazCheckbox-1\\" class=\\"m-switch\\" role=\\"switch\\" aria-checked=\\"false\\" tabindex=\\"0\\"><input id=\\"MazCheckbox-1\\" type=\\"checkbox\\" class=\\"m-switch__input\\"><span class=\\"m-switch__toggle\\"></span>
  <!--v-if-->
</label>"
`;
